File: app/src/main/java/com/jlbc/livedrawingjava/ParticleSystem.java
package com.jlbc.livedrawingjava;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PointF;
import java.util.ArrayList;
import java.util.Random;

/*
 *  This class represents a particle system.
 *
 *  PROPERTIES:
 *      running
 *      visible
 *
 *  METHODS
 *      init
 *      reset
 *      emit
 *      update
 *      draw
 */
public class ParticleSystem {

    // <editor-fold desc="INTERNAL FIELDS">

    private Random random = new Random();
    private ArrayList<Particle> particles = new ArrayList<>();
    private float duration = 0f;
    private boolean running = false;
    private boolean visible = false;

    // </editor-fold>

    // <editor-fold desc="PUBLIC PROPERTIES">

    public boolean getRunning() {
        return running;
    }

    public boolean getVisible() {
        return visible;
    }

    // </editor-fold>

    // <editor-fold desc="PUBLIC METHODS">

    // This method creates the particles of the system.
    public void init(int numParticles) {
        // Sets the condition of the system
        running = false;
        // Clears all particles that may exists
        particles.clear();
        for (int i = 0; i < numParticles; i++) {
            // Creates a random velocity vector (random angle in radians
            // and random magnitude between 1 and 10
            double angle = Math.toRadians(random.nextInt(360));
            double magnitude = random.nextInt(10) + 1;
            PointF velocity = new PointF(
                    (float)(Math.cos(angle) * magnitude),
                    (float)(Math.sin(angle) * magnitude));
            // Creates a particle and adds to the system
            particles.add(new Particle(velocity));
        }
    }

    // This method resets the particle system, basically
    // hides it and decrements the visible systems counter.
    public void reset(int numParticles) {
        if (visible) {
            visible = false;
            init(numParticles);
        }
    }

    // This method initiates the emission of particles
    // establishing all particles at same start position.
    public void emit(PointF startPosition) {
        running = true;
        visible = true;
        // Sets the duration of emission
        duration = 2f;
        // Sets the position of all particles
        for (Particle particle : particles) {
            particle.setPosition(startPosition);
        }
    }

    // This method updates all particles of the system and
    // verifies if system has reach its duration.
    public void update(long fps) {
        // Updates all particles
        for (Particle particle : particles) {
            particle.update(fps);
        }
        // Check if duration is reached
        duration -= (1f / fps);
        if (duration < 0) {
            running = false;
        }
    }

    // This method draws all particles in each frame. The system must
    // be visible.
    public void draw(Canvas canvas, Paint paint) {
        if (visible) {
            for (Particle particle : particles) {
                // Generates a random color to paint
                paint.setARGB(255,
                        random.nextInt(256),
                        random.nextInt(256),
                        random.nextInt(256));
                // Establishes the size of the particles
                float sizeX = 10;
                float sizeY = 10;
                // Draws the particle as a square
                canvas.drawRect(
                        particle.getPosition().x,
                        particle.getPosition().y,
                        particle.getPosition().x + sizeX,
                        particle.getPosition().y + sizeY,
                        paint);
            }
        }
    }

    // </editor-fold>
}

/*
 *  Reference: Horton (2018). Android Programming for Beginners
 */

File: app/src/main/java/com/jlbc/livedrawingkotlin/ParticleSystem.kt
package com.jlbc.livedrawingkotlin

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PointF
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/*
 *  This class represents a particle system.
 *
 *  PROPERTIES:
 *      running
 *      visible
 *
 *  METHODS
 *      init
 *      emit
 *      update
 *      draw
 */
class ParticleSystem {

    // <editor-fold desc="INTERNAL FIELDS">

    private val particles = ArrayList<Particle>()
    private var duration = 0f

    // </editor-fold>

    // <editor-fold desc="PUBLIC PROPERTIES">

    var running = false
    var visible = false

    // </editor-fold>

    // <editor-fold desc="PUBLIC METHODS">

    // This method creates the particles of the system.
    fun init(numParticles: Int) {
        // Sets the condition of the system
        running = false
        // Clears all particles that may exists
        particles.clear()
        for (i in 1..numParticles) {
            // Creates a random velocity vector (random angle in radians
            // and random magnitude between 1 and 10
            val angle = Random.nextInt(360) * Math.PI / 180
            val magnitude = Random.nextInt(1, 10)
            val velocity = PointF(
                (cos(angle) * magnitude).toFloat(),
                (sin(angle) * magnitude).toFloat())
            // Creates a particle and adds to the system
            particles.add(Particle(velocity))
        }
    }

    // This method resets the particle system, basically
    // hides it and decrements the visible systems counter.
    fun reset(numParticles: Int) {
        if (visible) {
            visible = false
            init(numParticles)
        }
    }

    // This method initiates the emission of particles
    // establishing all particles at same start position.
    fun emit(startPosition: PointF) {
        running = true
        visible = true
        // Sets the duration of emission
        duration = 2f
        // Sets the position of all particles
        particles.forEach {
            // IMPORTANT: In Java version, the setter "copies" the
            // components x,y of the parameters. Here in Kotlin, to
            // avoid use the same object reference, we need to create
            // a new PointF object for each particle.
            it.position = PointF(startPosition.x, startPosition.y)
        }
    }

    // This method updates all particles of the system and
    // verifies if system has reach its duration.
    fun update(fps: Long) {
        // Updates all particles
        particles.forEach {
            it.update(fps)
        }
        // Check if duration is reached
        duration -= (1f / fps)
        if (duration < 0) {
            running = false
        }
    }

    // This method draws all particles in each frame. The system must
    // be visible.
    fun draw(canvas: Canvas, paint: Paint) {
        if (visible) {
            particles.forEach {
                // Generates a random color to paint
                paint.setARGB(255,
                    Random.nextInt(256),
                    Random.nextInt(256),
                    Random.nextInt(256))
                // Establishes the size of the particles
                val sizeX = 10f
                val sizeY = 10f
                // Draws the particle as a square
                canvas.drawRect(
                    it.position.x,
                    it.position.y,
                    it.position.x + sizeX,
                    it.position.y + sizeY,
                    paint)
            }
        }
    }

    // </editor-fold>
}

/*
 *  Reference: Horton (2018). Android Programming for Beginners
 */